exts/virtio_block/lib.rs

```diff
@@ -0,0 +1,314 @@
+    size_max: u32,
+    seg_max: u32,
+    cylinders: u16,
+    heads: u8,
+    sectors: u8,
+    block_size: u32,
+    physical_block_exp: u8,
+    alignment_offset: u8,
+    min_io_size: u16,
+    opt_io_size: u32,
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C, packed)]
+struct VirtioBlockRequest {
+    type_: u32,
+    reserved: u32,
+    sector: u64,
+}
+
+#[repr(u32)]
+enum RequestType {
+    Read = 0,
+    Write = 1,
+}
+
+pub struct VirtioBlock {
+    virtio: Virtio,
+    request_buffer: VAddr,
+    status_buffer: VAddr,
+}
+
+impl VirtioBlock {
+    pub fn new(transport: Arc<dyn VirtioTransport>) -> Result<VirtioBlock, VirtioAttachError> {
+        let mut virtio = Virtio::new(transport);
+        virtio.initialize(VIRTIO_BLK_F_SIZE, 1)?;
+        // Read the block size
+        let block_size =
+            virtio.read_device_config64(offset_of!(VirtioBlockConfig, capacity) as u16);
+
+        info!("Block size is {} bytes", block_size);
+
+        let ring_len = virtio.virtq(VIRTIO_REQUEST_QUEUE).num_descs() as usize;
+
+        let request_buffer = alloc_pages(
+            (align_up(MAX_BLK_SIZE * ring_len, PAGE_SIZE)) / PAGE_SIZE,
+            AllocPageFlags::KERNEL,
+        )
+        .unwrap()
+        .as_vaddr();
+
+        let status_buffer = alloc_pages(
+            (align_up((MAX_BLK_SIZE * ring_len) + 1, PAGE_SIZE)) / PAGE_SIZE,
+            AllocPageFlags::KERNEL,
+        )
+        .unwrap()
+        .as_vaddr();
+
+        Ok(VirtioBlock {
+            virtio: virtio,
+            request_buffer: request_buffer,
+            status_buffer: status_buffer,
+        })
+    }
+
+    fn request_to_device(&mut self, request_type: RequestType, sector: u64, frame: &[u8]) {
+        let request_addr = self.request_buffer.add(MAX_BLK_SIZE);
+        let status_addr = self.status_buffer.add(MAX_BLK_SIZE + 1);
+        let buffer_len = size_of::<VirtioBlockRequest>();
+        let read_addr = unsafe { PAddr::new(*frame.as_ptr() as usize) };
+
+        // Fill block request
+        let block_request = unsafe { &mut *request_addr.as_mut_ptr::<VirtioBlockRequest>() };
+        block_request.type_ = request_type as u32;
+        block_request.sector = sector;
+        block_request.reserved = 0;
+
+        // Copy data into buffers
+        unsafe {
+            request_addr
+                .as_mut_ptr::<u8>()
+                .add(buffer_len)
+                .copy_from_nonoverlapping(frame.as_ptr(), frame.len());
+
+            status_addr
+                .as_mut_ptr::<u8>()
+                .add(buffer_len)
+                .copy_from_nonoverlapping(frame.as_ptr(), frame.len());
+        }
+
+        // Chain Descriptor
+        let chain = &[
+            VirtqDescBuffer::ReadOnlyFromDevice {
+                addr: request_addr.as_paddr(),
+                len: buffer_len + frame.len(),
+            },
+            VirtqDescBuffer::ReadOnlyFromDevice {
+                addr: read_addr,
+                len: buffer_len,
```

Comment:
buffer_len holds an incorrect size: it points to the size of the request header, not the size of payload.